clang/lib/Sema/SemaDecl.cpp

```diff
@@ -4049,6 +4029,13 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
       }
     }
 
+    // HLSL check parameters for matching ABI specifications.
+    if (getLangOpts().HLSL) {
+      if (HLSL().CheckCompatibleParameterABI(New, Old))
+        return true;
+      return MergeCompatibleFunctionDecls(New, Old, S, MergeTypeWithOld);
```

Comment:
`return MergeCompatibleFunctionDecls();` doesn't look right... I suspect this breaks the check for conflicting types.

Maybe we need to change hasSameFunctionTypeIgnoringExceptionSpec to hasSameFunctionTypeIgnoringExceptionSpecAndParameterABI?

Reply:
I wasn't entirely sure about that myself, but couldn't come up with a test case that illustrated a breakage. It is probably a failing of imagination on my part.

`hasSameFunctionTypeIgnoringExceptionSpec` is used in a bunch of places where I _think_ ignoring the parameter ABI would be a problem. Probably not for HLSL because HLSL doesn't support exceptions, but maybe for other cases that use parameter ABIs (swift interop).

I'll put up a patch shortly that brings back the `ASTContext::hasSameFunctionTypeIgnoringParamABI`, and uses that in conjunction with `hasSameFunctionTypeIgnoringExceptionSpec` here to be conservative. Does that sound reasonable?